rotkehlchen/chain/ethereum/modules/lido/decoder.py

```diff
@@ -0,0 +1,526 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any
+
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.constants import ZERO_ADDRESS
+from rotkehlchen.chain.evm.decoding.constants import ERC20_OR_ERC721_TRANSFER
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    ActionItem,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.constants.assets import A_ETH, A_STETH, A_WSTETH
+from rotkehlchen.fval import FVal
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from .constants import (
+    CPT_LIDO_ETH,
+    STETH_MAX_ROUND_ERROR_WEI,
+    STETH_SUBMIT_TOPIC,
+    STETH_TRANSFER_SHARES_TOPIC,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+steth_evm_address = A_STETH.resolve_to_evm_token().evm_address
+wsteth_evm_address = A_WSTETH.resolve_to_evm_token().evm_address
+
+
+class LidoDecoder(DecoderInterface):
+
+    def _decode_lido_staking_in_steth(self, context: DecoderContext) -> DecodingOutput:
+        """Decode the submit of eth to lido contract for obtaining steth in return"""
+
+        sender_address = hex_or_bytes_to_address(context.tx_log.topics[1])
+        amount_raw = hex_or_bytes_to_int(context.tx_log.data[:32])
+        collateral_amount = asset_normalized_value(
+            amount=amount_raw,
+            asset=A_ETH.resolve_to_crypto_asset(),
+        )
+
+        # Searching for the exact paired stETH token reception for validating the decoding
+        steth_minted_tokens = None
+        # Here searching for the paired stETH mint operation, with an hacky way for the issue:
+        # https://github.com/lidofinance/lido-dao/issues/442
+        for tx_log in context.all_logs:
```

Comment:
If the to is the sender address it will be seen as an ERC20 transfer by rotki. Why do you need to do this here. This does not make sense to me.

Reply:
I added the searching of tx_log for 2 reasons:
1. stETH ERC20 transfer decoded event is not yet available at this point, so we need to add an ActionItem as done below.
2. The expected amount of stETH tokens is not exactly known a this point, due to the lido rounding issue: https://github.com/lidofinance/lido-dao/issues/442. Thus, I am iterating the tx log for retreiving the exact stETH amount to be passed to the ActionItem "_amount=steth_minted_tokens_", using a custom STETH_MAX_ROUND_ERROR_WEI for finding the exact amount in tx logs (set as 8 for now, but we need to decide the final value, maybe checking with Lido team).